src/merkle/path.rs

```diff
@@ -23,14 +23,15 @@ impl MerklePath {
     // --------------------------------------------------------------------------------------------
 
     /// Computes the merkle root for this opening.
-    pub fn compute_root(&self, index_value: u64, node: Word) -> Word {
-        let mut index = NodeIndex::new(self.depth(), index_value);
-        self.nodes.iter().copied().fold(node, |node, sibling| {
+    pub fn compute_root(&self, index_value: u64, node: Word) -> Result<Word, MerkleError> {
+        let mut index = NodeIndex::new(self.depth(), index_value)?;
```

Comment:
This is an undesirable side effect of making a scalar index fallible. The computation of a root for a path should be infallible, and we should just use whatever the user input was as reference. This would give flexibility for the user to correctly compute roots for partial paths.

Reply:
I'm not sure I follow, how can this be infallible if the user can provide an index that is not present in the tree?

Do you want a `_unchecked` version of this?

Reply:
One example is verifying openings for intersecting paths. Suppose we have the following tree

![image](https://user-images.githubusercontent.com/8730839/224777708-04192806-e74f-4f69-8904-7ca54c6ac4dd.png)

The index of the blue leaf is `0b0000`, and the index of the green leaf is `0b1100`.

In order to verify the opening for both leaves, we can naively perform a full check on both openings to the root. However, a good optimization there is to check the opening from the green leaf to the intersecting orange node, and then the opening from the blue leaf to the tree root.

To do that, if we don't perform such validation, we can just truncate the merkle path/depth and use the indexes. The green path to orange would work both with `0b1100` and `0b11`, even tho `0b1100` is invalid for depth `2`.

This is ofc just one example, but this optimization wouldn't be possible if we asserted the depth correctness here. The trade-off we need to keep in mind is restricting potential usage vs safety. I'm not sure we are improving our safety by checking the depth vs index relationship at every tick; OTOH, we can think of a couple of examples that we are making unfeasible if we introduce such check.

Reply:
The problem there is we force the user to manipulate the bits of the index when this responsibility should be of the `NodeIndex`.

Creating `checked`/`unchecked` methods everywhere will double our API complexity. I think it should be used in some cherry-picked cases, and not as a broad approach

Reply:
> The problem there is we force the user to manipulate the bits of the index when this responsibility should be of the NodeIndex.

How is that different from your proposal?

> To do that, if we don't perform such validation, we can just truncate the merkle path/depth and use the indexes.

I would also just make that into some API, e.g. `change_root`

> Creating checked/unchecked methods everywhere will double our API complexity. I think it should be used in some cherry-picked cases, and not as a broad approach

Totally agree, that is why I have been insisting to only introduce `unchecked` after we benchmark and show that the method will have a practical impact